src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/SimpleTypeInfos.cs

```diff
@@ -292,9 +292,12 @@ public override void WriteData(PropertyValue value)
         {
             // It's not currently possible to get the HasValue property of a nullable type through reflection when the
             // value is null. Instead, we simply check that the nullable is not null.
-            bool hasValue = value.ReferenceValue != null;
+            object? refVal = value.ReferenceValue;
+            bool hasValue = refVal is not null;
             TraceLoggingDataCollector.AddScalar(hasValue);
-            PropertyValue val = hasValue ? valueGetter(value) : valueInfo.PropertyValueFactory(Activator.CreateInstance(valueInfo.DataType));
+            PropertyValue val = valueInfo.PropertyValueFactory(hasValue
+                ? value
+                : Activator.CreateInstance(valueInfo.DataType));
```

Comment:
I don't think `Activator.CreateInstance` is correct here. I've been fixing these (or logging issues) as I've encountered them. If this was `Nullable<SomeStruct>` where `SomeStruct` had a parameterless ctor (https://github.com/dotnet/csharplang/issues/99), this would call the `ctor`, which isn't correct.

Instead, this should call `RuntimeHelpers.GetUninitializedObject(valueInfo.DataType)`, and then we should suppress the warning in code, because it is always safe to initialize a ValueType. For example:

https://github.com/dotnet/runtime/blob/5d5c3e7a589e21e43ad6464b6ed6732e98e7fc8a/src/libraries/Common/src/Extensions/ParameterDefaultValue/ParameterDefaultValue.cs#L56-L63

cc @MichalStrehovsky

Reply:
I'm not sure I follow the suppression logic. What is stopping `SomeStruct` from being trimmed just because it's a value type? Wouldn't be that `SomeStruct` couldn't be trimmed out because creating a `Nullable<SomeStruct>` requires `SomeStruct` to be preserved for type identity?

Reply:
`valueInfo.DataType` will be `typeof(SomeStruct)`. Since someone is calling into this method, `SomeStruct` can't be trimmed. And since this code is only used for `Nullable<T>`, you know that `T` is a ValueType.

As for the suppression, in my sample above we are suppressing the warning that the following annotation is not being met:

https://github.com/dotnet/runtime/blob/8a52f1e948b6f22f418817ec1068f07b8dae2aa5/src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs#L148-L154

Since we know that this is a `ValueType`, we know that we can always call `GetUninitializedObject` on it.

Reply:
This makes sense, but it seems like the key is that this method won't be called unless someone instantiated `SomeStruct` at some point. The phrase "you can always create an instance of a valuetype" seems wrong just because without the previous precondition it seems possible that it could fail.

Reply:
Yeah, maybe it's better to use the "valuetypes are considered implicitly boxed during trimming". (The annotation on GetUninitializedObject just needs to make sure illink doesn't consider the type never allocated on the GC heap and go removing virtual overrides and stubbing out interface implementations.)